File: src/jarvis/tools/shell.py
"""Universal shell execution tool for Jarvis.

Enables hands-free execution of arbitrary Linux commands and shell scripts.
"""

import asyncio
import os
from typing import Optional
from jarvis.core.logger import log


class ShellExecutor:
    """Safely executes shell commands on Omarchy Linux with output truncation and timeouts."""

    def __init__(self, default_timeout: float = 15.0, max_output_chars: int = 2000):
        self.default_timeout = default_timeout
        self.max_output_chars = max_output_chars

    async def execute_command(self, command: str, timeout: Optional[float] = None) -> str:
        """Execute an arbitrary bash command and return its output.

        Args:
            command: The bash command string to execute.
            timeout: Maximum execution time in seconds before terminating (default: 15.0s).
        """
        if not command or not command.strip():
            return "No command provided."

        cmd_clean = command.strip()
        actual_timeout = timeout if timeout is not None and timeout > 0 else self.default_timeout
        log.info(f"[cyan]Executing shell command (timeout={actual_timeout}s): {cmd_clean}[/cyan]")

        try:
            proc = await asyncio.create_subprocess_shell(
                cmd_clean,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                executable="/bin/bash",
                env=os.environ.copy(),
            )

            try:
                stdout_bytes, stderr_bytes = await asyncio.wait_for(
                    proc.communicate(),
                    timeout=actual_timeout,
                )
            except asyncio.TimeoutError:
                try:
                    proc.kill()
                    await proc.wait()
                except Exception:
                    pass
                log.warning(f"Shell command timed out after {actual_timeout}s: {cmd_clean}")
                return f"Command timed out after {actual_timeout} seconds."

            stdout = stdout_bytes.decode("utf-8", errors="replace").strip()
            stderr = stderr_bytes.decode("utf-8", errors="replace").strip()
            return_code = proc.returncode

            # Format result
            output_parts = []
            if stdout:
                output_parts.append(stdout)
            if stderr:
                output_parts.append(f"[stderr]:\n{stderr}")

            combined = "\n".join(output_parts).strip()

            if not combined:
                if return_code == 0:
                    return "Command executed successfully (no output)."
                else:
                    return f"Command exited with return code {return_code} (no output)."

            # Truncate if exceeds max_output_chars
            if len(combined) > self.max_output_chars:
                half = self.max_output_chars // 2
                truncated = (
                    combined[:half]
                    + f"\n\n... [output truncated: {len(combined)} characters total] ...\n\n"
                    + combined[-half:]
                )
                return truncated

            return combined

        except Exception as e:
            log.error(f"Failed to execute shell command '{cmd_clean}': {e}")
            return f"Execution error: {e}"

File: src/tools/shell.rs
use crate::core::error::{JarvisError, Result};
use crate::tools::Tool;
use async_trait::async_trait;
use serde_json::{json, Value};
use std::sync::Arc;
use std::time::Duration;
use tokio::process::Command;
use tokio::time::timeout;
use tracing::{debug, warn};

pub struct ShellExecutor {
    default_timeout: Duration,
    max_output_chars: usize,
}

impl Default for ShellExecutor {
    fn default() -> Self {
        Self::new(15.0, 2000)
    }
}

impl ShellExecutor {
    pub fn new(default_timeout_secs: f64, max_output_chars: usize) -> Self {
        Self {
            default_timeout: Duration::from_secs_f64(default_timeout_secs),
            max_output_chars,
        }
    }

    pub async fn execute_command(
        &self,
        command: &str,
        custom_timeout: Option<f64>,
    ) -> Result<String> {
        let trimmed = command.trim();
        if trimmed.is_empty() {
            return Ok("No command provided.".to_string());
        }

        let run_timeout = custom_timeout
            .map(Duration::from_secs_f64)
            .unwrap_or(self.default_timeout);

        debug!(
            "Executing shell command (timeout={:?}): {}",
            run_timeout, trimmed
        );

        let mut child = Command::new("/bin/bash")
            .arg("-c")
            .arg(trimmed)
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn()?;

        let mut stdout_pipe = child.stdout.take().unwrap();
        let mut stderr_pipe = child.stderr.take().unwrap();

        let read_future = async {
            let mut stdout_buf = Vec::new();
            let mut stderr_buf = Vec::new();
            let (_, _, status) = tokio::join!(
                tokio::io::AsyncReadExt::read_to_end(&mut stdout_pipe, &mut stdout_buf),
                tokio::io::AsyncReadExt::read_to_end(&mut stderr_pipe, &mut stderr_buf),
                child.wait()
            );
            (stdout_buf, stderr_buf, status)
        };

        let result = timeout(run_timeout, read_future).await;

        match result {
            Ok((stdout_bytes, stderr_bytes, Ok(status))) => {
                let stdout = String::from_utf8_lossy(&stdout_bytes).trim().to_string();
                let stderr = String::from_utf8_lossy(&stderr_bytes).trim().to_string();
                let code = status.code().unwrap_or(-1);

                let mut parts = Vec::new();
                if !stdout.is_empty() {
                    parts.push(stdout);
                }
                if !stderr.is_empty() {
                    parts.push(format!("[stderr]:\n{stderr}"));
                }

                let combined = parts.join("\n").trim().to_string();

                if combined.is_empty() {
                    if status.success() {
                        return Ok("Command executed successfully (no output).".to_string());
                    } else {
                        return Ok(format!(
                            "Command exited with return code {code} (no output)."
                        ));
                    }
                }

                if combined.len() > self.max_output_chars {
                    let half = self.max_output_chars / 2;
                    let truncated = format!(
                        "{}\n\n... [output truncated: {} characters total] ...\n\n{}",
                        &combined[..half],
                        combined.len(),
                        &combined[combined.len() - half..]
                    );
                    return Ok(truncated);
                }

                Ok(combined)
            }
            Ok((_, _, Err(e))) => Err(JarvisError::Io(e)),
            Err(_) => {
                let _ = child.kill().await;
                warn!("Shell command timed out: {trimmed}");
                Ok(format!(
                    "Command timed out after {} seconds.",
                    run_timeout.as_secs()
                ))
            }
        }
    }
}

// -----------------------------------------------------------------------------
// Tool Implementations
// -----------------------------------------------------------------------------

pub struct ExecuteCommandTool {
    executor: Arc<ShellExecutor>,
}

impl ExecuteCommandTool {
    pub fn new(executor: Arc<ShellExecutor>) -> Self {
        Self { executor }
    }
}

#[async_trait]
impl Tool for ExecuteCommandTool {
    fn name(&self) -> &'static str {
        "execute_command"
    }

    fn description(&self) -> &'static str {
        "Execute any bash command directly on the Linux system and return its terminal output."
    }

    fn parameters_schema(&self) -> Value {
        json!({
            "type": "OBJECT",
            "properties": {
                "command": {
                    "type": "STRING",
                    "description": "The exact bash command line to run."
                },
                "timeout": {
                    "type": "NUMBER",
                    "description": "Optional timeout in seconds before aborting (default: 15.0)."
                }
            },
            "required": ["command"]
        })
    }

    async fn execute(&self, args: Value) -> Result<String> {
        let command = args["command"].as_str().ok_or_else(|| {
            JarvisError::ToolParameter(
                "execute_command".into(),
                "command string is required".into(),
            )
        })?;
        let timeout_secs = args["timeout"].as_f64();

        self.executor.execute_command(command, timeout_secs).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_shell_executor_echo() {
        let executor = ShellExecutor::default();
        let out = executor
            .execute_command("echo 'jarvis test'", None)
            .await
            .unwrap();
        assert_eq!(out, "jarvis test");
    }

    #[tokio::test]
    async fn test_shell_executor_timeout() {
        let executor = ShellExecutor::new(0.5, 500);
        let out = executor.execute_command("sleep 2", None).await.unwrap();
        assert!(out.contains("timed out"));
    }
}
